File: Numeros Pseudo-Aleatorios/Permuted Congruential Generator/pcg.c
/*
 * PRNG .......: Permuted Congruential Generator (PCG)
 * Ano ........: 2014
 * Autor ......: Dr. M.E. O'Neill
 */

#include <stdio.h>
#include <stdint.h>
#include <time.h>

static uint32_t rotr32(uint32_t x, unsigned r)
{
  return x >> r | x << (-r & 31);
}

uint32_t pcg32(uint64_t* seed)
{
  uint64_t x = *seed;
  unsigned rotation = (unsigned)(x >> 59);
  *seed = x * 6364136223846793005u + 1442695040888963407u;
  x ^= x >> 18;
  *seed = *seed + x;
  return rotr32((uint32_t)(((x >> 27) + rotation) ^ (*seed >> 32)), rotation);
}

uint32_t getTimer()
{
  time_t current_time = time(NULL);
  return (uint32_t)current_time;
}

uint64_t Xn1, Xn2;

int main(void)
{

  Xn1 = 1234;
  Xn2 = getTimer();
  
  printf("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)\n");
  printf("PSEUDORANDOM NUMBER GENERATOR (PRNG)\n");
  printf("PADRAO: Permuted Congruential Generator (PCG)\n");
  printf("=============================================\n");
  printf("\n");
  printf("      Semente: 1234       Semente: TIME\n");
  printf("      -------------       -------------\n");

  for (int i = 1; i <= 10; i++)
    printf("%19u%20u\n", pcg32(&Xn1), pcg32(&Xn2));

  return 0;
}

File: Numeros Pseudo-Aleatorios/Permuted Congruential Generator/pcg.cs
/*
 * PRNG .......: Permuted Congruential Generator (PCG)
 * Ano ........: 2014
 * Autor ......: Dr. M.E. O'Neill
 */

using System;
using System.Diagnostics;

public class Program
{

    private static uint rotr32(uint x, int r)
    {
        return (x >> r) | (x << (-r & 31));
    }

    private static uint pcg32(ref long seed)
    {
        long x = seed;
        int rotation = (int)(x >> 59);
        seed = x * 6364136223846793005L + 1442695040888963407L;
        x ^= x >> 18;
        seed += x;
        return rotr32((uint)(((x >> 27) + rotation) ^ (seed >> 32)), rotation);
    }

    private static uint GetTimer()
    {
        var current_time = DateTime.UtcNow;
        var millis = (long)(current_time - new DateTime(1970, 1, 1)).TotalMilliseconds;
        return (uint)millis;
    }

    private static long Xn1, Xn2;

    public static void Main(string[] args)
    {
        Console.WriteLine("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)");
        Console.WriteLine("PSEUDORANDOM NUMBER GENERATOR (PRNG)");
        Console.WriteLine("PADRAO: Permuted Congruential Generator (PCG)");
        Console.WriteLine("=============================================");
        Console.WriteLine();

        Xn1 = 1234;
        Xn2 = GetTimer();

        Console.WriteLine("      Semente: 1234       Semente: TIME");
        Console.WriteLine("      -------------       -------------");

        for (int i = 1; i <= 10; i++)
        {
            Console.Write($"{pcg32(ref Xn1),19}");
            Console.WriteLine($"{pcg32(ref Xn2),20}");
        }
    }
}
